lib/leaf.go

```diff
@@ -0,0 +1,131 @@
+package lib
+
+import (
+	"reflect"
+)
+
+const (
+	tagName             = "autumn"
+	getNameMethod       = "GetLeafName"
+	postConstructMethod = "PostConstruct"
+)
+
+// Leaf describes a single injected class
+type Leaf struct {
+	structureType    reflect.Type
+	structureValue   reflect.Value
+	structureElement reflect.Value
+
+	name                   string
+	value                  interface{}
+	unresolvedDependencies map[string]reflect.Value
+	resolvedDependencies   map[string]reflect.Value
+}
+
+// NewLeaf constructs a new leaf, using the structure name as the name
+func NewLeaf(structPtr interface{}) *Leaf {
+	leaf := &Leaf{
+		structureType:    getStructureType(structPtr),
+		structureValue:   getStructureValue(structPtr),
+		structureElement: getStructureElement(structPtr),
+	}
+
+	leaf.initializeName()
+	leaf.initializeDependencies()
+	return leaf
+}
+
+// NewNamedLeaf constructs a new leaf with the specified name
+func NewNamedLeaf(name string, structPtr interface{}) *Leaf {
+	leaf := &Leaf{
+		structureType:    getStructureType(structPtr),
+		structureValue:   getStructureValue(structPtr),
+		structureElement: getStructureElement(structPtr),
+		name:             name,
+	}
+
+	leaf.initializeDependencies()
+	return leaf
+}
+
+// initializeName initializes the name for the leaf
+func (l *Leaf) initializeName() {
+
+	method := l.structureValue.MethodByName(getNameMethod)
+	if !method.IsValid() {
+		l.name = l.structureType.String()
+		return
+	}
+
+	if method.Type().NumIn() != 0 {
+		panic(l.structureType.String() + " - " + getNameMethod + " must not take any parameters")
+	} else if method.Type().NumOut() != 1 {
+		panic(l.structureType.String() + " - " + getNameMethod + " must return exactly one parameter")
+	} else if method.Type().Out(0).Kind() != reflect.String {
+		panic(l.structureType.String() + " - " + getNameMethod + " must return a string")
+	}
+
+	l.name = method.Call([]reflect.Value{})[0].String()
+}
+
+// initializeDependencies read in structure tags to find dependencies
+func (l *Leaf) initializeDependencies() {
+	l.unresolvedDependencies = map[string]reflect.Value{}
+	l.resolvedDependencies = map[string]reflect.Value{}
+
+	for i := 0; i < l.structureType.NumField(); i++ {
```

Comment:
Looks like you only loop through the first level of dependencies? What if A has a dependency on B, and then B has a dependency on C?

Reply:
If A has a dependency on B, the pointer to B will be injected in A. Once B is resolved, the pointer for C will be stuck into that B, which means that A.B.C will work once all deps have been resolved.